File: app/src/types/boe-packing.ts
export type BoePackStage =
  | "BOE_PACK_SCAN_PLAN"
  | "BOE_PACK_FETCH_WMS"
  | "BOE_PACK_ENRICH"
  | "BOE_PACK_SAVE_DRAFT"
  | "BOE_PACK_REVIEW"
  | "BOE_PACK_SUBMITTING"
  | "BOE_PACK_SUBMITTED"
  | "BOE_PACK_CANCELLED";

export type BoePackHeader = {
  aiRecognize?: boolean;
  invoiceNo?: string;
  factory?: string;
  customerName?: string;
  customerSubcode?: string;
  businessEntity?: string;
  invoiceDate?: string;
  etd?: string;
  consignArrivalDate?: string;
  totalVol?: string;
  volUnit?: string;
};

export type BoePackAttachType = "箱单" | "发票" | "提运单" | "双签PO/协议";

export type BoePackAttachment = {
  id?: string;
  type?: BoePackAttachType | string;
  fileName?: string;
  filePath?: string;
};

export type BoePackLine = {
  lineNo?: string;
  poNum?: string;
  itemNum?: string;
  deliveryQty?: string;
  netWeight?: string;
  netWeightUnit?: string;
  regionCode?: string;
  regionSrmName?: string;
  lineItem?: string;
  orderQty?: string;
  orderUnit?: string;
  remainingQty?: string;
  itemName?: string;
  factory?: string;
};

export type BoePackListItem = {
  id: string;
  processCode: string;
  bizKey: string;
  title: string;
  portalAccountId: string;
  stage: BoePackStage;
  status: string;
  lineTotal: number;
  lineDone: number;
  lastErrorCode?: string | null;
  lastErrorMessage?: string | null;
  createdAt: string;
  updatedAt: string;
  qtyMismatch?: boolean;
  invoiceNo?: string;
  factory?: string;
  customerName?: string;
  srmDraftNo?: string;
  headerId?: string;
};

export type BoePackDetail = BoePackListItem & {
  header: BoePackHeader;
  lines: BoePackLine[];
  qtyWarning?: string | null;
  orgCodeWarning?: string | null;
  reviewBaseline?: Record<string, unknown> | null;
  attachments?: BoePackAttachment[];
  stageHistory: Array<{
    id: string;
    fromStage?: string | null;
    toStage: string;
    actor: string;
    note?: string | null;
    createdAt: string;
  }>;
  subTasks: Array<{
    id: string;
    title: string;
    taskType: string;
    status: string;
    createdAt: string;
    updatedAt: string;
    lineNumber?: string | null;
  }>;
};

export type BoeMatchResult = {
  createdCount: number;
  skippedCount: number;
  missingPortal: string[];
  error?: string | null;
  createdIds: string[];
};

File: service/app/schemas/boe_packing.py
from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.common import CamelModel


class BoePackingListItem(CamelModel):
    id: str
    process_code: str = Field(serialization_alias="processCode")
    biz_key: str = Field(serialization_alias="bizKey")
    title: str
    portal_account_id: str = Field(serialization_alias="portalAccountId")
    stage: str
    status: str
    line_total: int = Field(serialization_alias="lineTotal")
    line_done: int = Field(serialization_alias="lineDone")
    last_error_code: str | None = Field(None, serialization_alias="lastErrorCode")
    last_error_message: str | None = Field(None, serialization_alias="lastErrorMessage")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")
    qty_mismatch: bool = Field(False, serialization_alias="qtyMismatch")
    invoice_no: str = Field("", serialization_alias="invoiceNo")
    factory: str = ""
    customer_name: str = Field("", serialization_alias="customerName")
    srm_draft_no: str = Field("", serialization_alias="srmDraftNo")
    header_id: str = Field("", serialization_alias="headerId")


class BoePackingHeaderPatch(CamelModel):
    invoice_no: str | None = Field(None, alias="invoiceNo")
    factory: str | None = None
    invoice_date: str | None = Field(None, alias="invoiceDate")
    etd: str | None = None
    consign_arrival_date: str | None = Field(None, alias="consignArrivalDate")
    total_vol: str | None = Field(None, alias="totalVol")


class BoePackingLinePatch(CamelModel):
    line_no: str | None = Field(None, alias="lineNo")
    po_num: str | None = Field(None, alias="poNum")
    item_num: str | None = Field(None, alias="itemNum")
    delivery_qty: str | None = Field(None, alias="deliveryQty")
    net_weight: str | None = Field(None, alias="netWeight")
    net_weight_unit: str | None = Field(None, alias="netWeightUnit")
    region_code: str | None = Field(None, alias="regionCode")
    region_srm_name: str | None = Field(None, alias="regionSrmName")
    line_item: str | None = Field(None, alias="lineItem")
    order_qty: str | None = Field(None, alias="orderQty")
    order_unit: str | None = Field(None, alias="orderUnit")
    remaining_qty: str | None = Field(None, alias="remainingQty")
    item_name: str | None = Field(None, alias="itemName")
    factory: str | None = None


class BoePackingAttachmentPatch(CamelModel):
    id: str | None = None
    type: str | None = None
    file_name: str | None = Field(None, alias="fileName")
    file_path: str | None = Field(None, alias="filePath")


class BoePackingPatchRequest(CamelModel):
    header: BoePackingHeaderPatch | None = None
    lines: list[BoePackingLinePatch] | None = None
    attachments: list[BoePackingAttachmentPatch] | None = None


class BoeMatchResponse(CamelModel):
    created_count: int = Field(serialization_alias="createdCount")
    skipped_count: int = Field(serialization_alias="skippedCount")
    missing_portal: list[str] = Field(default_factory=list, serialization_alias="missingPortal")
    error: str | None = None
    created_ids: list[str] = Field(default_factory=list, serialization_alias="createdIds")
    skipped: list[dict[str, Any]] = Field(default_factory=list)
